client/lib/wpcom-undocumented/lib/undocumented.js

```diff
@@ -817,9 +824,38 @@ Undocumented.prototype.cart = function( cartKey, method, data, fn ) {
 		method = 'GET';
 		data = {};
 	}
+
+	if ( method === 'GET' && cartRequestCounter === 1 ) {
+		// If it's the first GET request and we have `?coupon=[code]`, then
```

Comment:
Modifying the cart with GET API calls contradicts the principles of REST :(

Reply:
@dzver Thanks for reviewing! :-)

I see your point and mostly agree.

On the other hand, Calypso makes an initial GET request to acquire the current state of the cart so it has something to work with. A `?coupon=[code]` in the GET request makes this return an alternate representation of the same data; i.e., the same cart contents, but discounted by the coupon code. I'm not sure I see the coupon being a resource itself. I see it as being a property of a resource (cart item prices).

We could make the initial GET for the current state of the cart, and then make an additional `POST` to apply the coupon. Do you feel it is worth the additional trip?

Reply:
It's not an alternate representation because the cart is persistent. 

Both the cookie and the server diff do not look necessary to me and I'll be happy if we can find a way without them.

Reply:
Thank you. I can produce a fork of this PR without the cookie and put the focus only on the query string. If I understand correctly, that's what you'd like to see. Do I have that right?

Having said that, I have some requests for this to work for affiliates too. So being able to pass `?coupon=[code]` into any Calypso endpoint and/or landing pages, save it as a cookie, and then use it at checkout would be helpful in my view. Do you see a better way to facilitate that?

Reply:
Yes, they can use the link to checkout or the affiliate landing pages can link to the checkout by providing the additional query string.

Reply:
Understood and can do. Thank you.